EncoreTickets.SDK.Tests/UnitTests/Pricing/PricingServiceApiTests.cs

```diff
@@ -0,0 +1,694 @@
+        {
+            TestSuccessfulAction(() => GetPartnersInGroup(1), expectedResult);
+        }
+
+        [Test]
+        public void GetPartnersInGroup_NotFoundGroup_ThrowsApiException()
+        {
+            TestApiExceptionOnFailedAction<IList<Partner>>(
+                () => GetPartnersInGroup(-1),
+                NotFoundResponseContent,
+                HttpStatusCode.NotFound);
+        }
+
+        [Test]
+        public void GetPartner_CorrectlyCalled()
+        {
+            const int id = 1;
+
+            TestActionCorrectlyExecuted<Partner>(
+                () => GetPartner(id),
+                $"v{ApiVersion}/admin/partners/{id}",
+                Method.GET,
+                shouldAuthenticate: true);
+        }
+
+        [TestCaseSource(typeof(PricingServiceApiTestSource), nameof(GetPartner_ReturnsPartner))]
+        public void GetPartner_ReturnsPartner(Partner expectedResult)
+        {
+            TestSuccessfulAction(() => GetPartner(1), expectedResult);
+        }
+
+        [Test]
+        public void GetPartner_NotFoundPartner_ThrowsApiException()
+        {
+            TestApiExceptionOnFailedAction<Partner>(
+                () => GetPartner(-1),
+                NotFoundResponseContent,
+                HttpStatusCode.NotFound);
+        }
+
+        private void TestActionCorrectlyExecuted<T>(Action action,
+            string endpoint,
+            Method method,
+            Dictionary<string, object> expectedQueryParameters = null,
+            bool shouldAuthenticate = false)
+            where T : class, new()
+        {
+            TestActionCorrectlyExecuted<T, T>(action, endpoint, method, expectedQueryParameters, shouldAuthenticate);
+        }
+
+        private void TestActionCorrectlyExecuted<TSetup, TResult>(Action action,
+            string endpoint,
+            Method method,
+            Dictionary<string, object> expectedQueryParameters = null,
+            bool shouldAuthenticate = false)
+            where TSetup : class, TResult, new()
+            where TResult : class
+        {
+            mocker.SetupAnyExecution<TSetup>();
+
+            try
+            {
+                action();
+            }
+            catch (Exception)
+            {
+                // ignored
+            }
+
+            if (shouldAuthenticate)
+            {
+                ShouldAuthenticate();
+            }
+            else
+            {
+                ShouldNotAuthenticate();
+            }
+            mocker.VerifyExecution<ApiResponse<TResult>>(BaseUrl,
+                endpoint, method, expectedQueryParameters: expectedQueryParameters);
+        }
+
+        private void TestSuccessfulAction<T>(Func<T> action, T expectedResult)
+            where T : class
+        {
+            var serializedExpectedResult = serializer.Serialize(expectedResult);
+            var responseContent = string.Format(SuccessfulResponseTemplate, serializedExpectedResult);
```

Comment:
I'm not sure I really like this method for writing response-based tests. You separate tests for serializing models from serializing the full response and do not test them together anywhere. But in the model wrapper there can be really useful information (for example, in the “context” field), on which the result of the tested method may depend. And here you are doing some additional logic just to prepare the source data, and such source data is calculated when I would rather see maybe large but raw data, the same data that we can get in response from the API. 
For me, setting a full response without calculating and setting mocks to simulate that the API really returns such a response is better. And at the same time, it’s really good to have tests to just check the serialization of models, that’s fine, thanks. But for testing of services, I prefer full response content for a test source.

Reply:
I don't see any reason why we can't simulate "really useful information (for example, in the “context” field)" with this approach. The implemented methods don't contain any useful information in the "context" or "request" fields if the response is successful, that's why there is no useful information in the corresponding fields in the template I use for successful responses. But if there were, there is nothing stopping us from simply creating another template. Since the data in those fields is irrelevant for what we are trying to test, it should not matter what we have there and I'd much rather have a single such template than have a wall of irrelevant text on every single TestCaseSource which nobody ever reads anyway and which takes up at least half the screen during code reviews.

At the same time the whole point of creating separate tests for model serialization was to make sure that serialization and deserialization is done correctly so that we could be confident that when we serialize the object in this class, we are actually getting the string we would get from the API and not something else. With that, I don't see why would we rather use raw responses: this just creates unnecessary redundancy and makes any kind of edits to the test data more painful since we have to make edits in two places instead of just one. I already had to make such edits to existing test cases and I would definitely take the new approach over the old one